### PR TITLE
Update model_graph.py to add model_to_networkx method

### PR DESCRIPTION
**Pull request by jonititan**
Created a new file based on the model_graph.py file to create a method for exporting the graph network structure to networkx.
This is useful as it allows direct conversion instead of exporting to graphviz, saving to dot file, importing the dotfile via pydot before finally creating the networkx graph.

The subgraph method is untested but the remainder works reliably so far.

This is a new method so there should be no breaking changes.
Docstrings have all been updated.
The same approach could be used in v4.  I did it in v3 because that's what I'm currently using.

**Thank your for opening a PR!**

Before you proceed, please make sure that [the pre-commit linting/style checks pass](https://docs.pymc.io/en/latest/contributing/python_style.html).

Depending on what your PR does, here are a few things you might want to address in the description:
+ [x] what are the (breaking) changes that this PR makes?
+ [x] important background, or details about the implementation
+ [x] are the changes—especially new features—covered by tests and docstrings?
+ [x] [consider adding/updating relevant example notebooks](https://github.com/pymc-devs/pymc-examples)
+ [ ] right before it's ready to merge, mention the PR in the RELEASE-NOTES.md

### CODE REVIEW

**Comment by twiecki** [comment]
Thanks for working on this! What's your use-case? Plotting?

**Comment by jonititan** [comment]
Yes and checking model structure on complex models.

Upon further reflection the cluster label section will not work correctly as graphviz .subgraph incorporates nodes added to it onto the graph network as a subgraph wheras in networkx the same method name creates a read only view.
It is difficult for me to test as I do not have any models with clusters.
I am currently rewriting that bit.

**Comment by jonititan** [comment]
I have updated the subgraph/cluster handling so that the information is preserved.  That should all work no problems now but I would appreciate if someone with a model that uses clusters/subgraphs could either test it or send me a link to their model code so I can test it myself.

**Comment by michaelosthege** [comment]
I like the idea of having more export options. But we should make sure that we're not duplicating code, also to keep testing simple.

There's another WIP PR related to the model graph: https://github.com/pymc-devs/pymc/pull/5677

Overall, the model graph would greatly benefit from a clearer definition of its attributes/properties and the extraction of methods like `get_plates` or `get_deterministics` into independent functions.

**Comment by jonititan** [comment]
I can see the logic there.

Given the minimal difference between exporting to graphviz and exporting to networkx that should be easy to support.

I just did it as a new file to avoid causing errors in existing methods.

**Comment by jonititan** [comment]
I have now modified model_graph.py to reflect those comments.
model_graph.py now works correctly to convert pymc3 v3 models to either graphviz or networkx representations though as noted none of my models feature clusters so there is not complete test coverage.

File: pymc3/model_graph.py [hunk 1]
```diff
@@ -142,21 +142,25 @@ def _make_node(self, var_name, graph, *, formatting: str = "plain"):
...
 
-        graph.node(var_name.replace(":", "&"), label, **attrs)
+        if nx:
+            graph.add_node(var_name.replace(":", "&"), label=label, **attrs)
+        else:
+            graph.node(var_name.replace(":", "&"), label, **attrs)
 
     def get_plates(self):
         """Rough but surprisingly accurate plate detection.
-
```
Comment (michaelosthege) [inline pymc3/model_graph.py:156]: Please undo the whitespace changes here and 159, 187, 224, 227, 232.
The empty line is part if the numpydoc style.

File: pymc3/model_graph.py [hunk 2]
```diff
@@ -218,18 +221,46 @@ def make_graph(self, formatting: str = "plain"):
...
+        -------
+        graphviz.Digraph
+        """
+        try:
+            import networkx
+        except ImportError:
+            raise ImportError(
+                "This function requires the python library graphviz, along with binaries. "
+                "The easiest way to install all of this is by running\n\n"
+                "\tconda install -c conda-forge python-graphviz"
```
Comment (michaelosthege) [inline pymc3/model_graph.py:236]:
```suggestion
                "This function requires networkx. "
                "The easiest way to install all of this is by running\n\n"
                "\tconda install -c conda-forge networkx"
```

**Comment by michaelosthege** [comment]
I like this! Already have an idea where I want to try out `pm.model_to_networkx`.

Before we can merge, we'll need a few tests. I'll also trigger approve running the CI then - at the moment it doesn't make sense without a test.

1. In [`requirements-dev.txt`](https://github.com/pymc-devs/pymc/blob/main/requirements-dev.txt) and all environment YAMLs under `/conda-envs`, add `networkx`
2. Add a test case by duplicating and adapting [`test_graphviz`](https://github.com/pymc-devs/pymc/blob/main/pymc/tests/test_model_graph.py#L163-L171)
3. Try the test locally with `pytest -v pymc\tests\test_model_graph.py -k test_networkx`
4. Run `pre-commit run --all` to apply code style.

**Comment by jonititan** [comment]
Glad to hear it.  Do you need anything else from me on this now to carry it through?

**Comment by michaelosthege** [comment]
> Glad to hear it. Do you need anything else from me on this now to carry it through?

I didn't check it out yet, but your branch looks very promising. Tomorrow I'll try the export with one of my models.

**Review by michaelosthege — COMMENTED** [review]
Hi @jonititan,
at first I didn't notice that your PR targets `v3`. We're no longer taking new features into `v3` because it adds maintenance burden. Our focus is on the upcoming `4.0.0` release from the `main` branch.

The model I wanted to test your function with is also built with `v4`, so I made a new branch from `main` where I applied your changes. See https://github.com/pymc-devs/pymc/pull/5705

I tried to export the graph of a small toy model to GraphML, but ran into this exception:
```
Traceback (most recent call last):
  File "C:\Users\osthege\AppData\Local\Continuum\miniconda3\envs\pm3v4\lib\site-packages\networkx\readwrite\graphml.py", line 455, in get_xml_type
    return self.xml_type[key]
KeyError: <class 'NoneType'>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "C:\Users\osthege\AppData\Local\Continuum\miniconda3\envs\pm3v4\lib\site-packages\IPython\core\interactiveshell.py", line 3369, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "C:\Users\osthege\AppData\Local\Temp\ipykernel_35436\1366585194.py", line 1, in <cell line: 1>
    networkx.readwrite.write_graphml(graph, "minimodel.gml")
  File "<class 'networkx.utils.decorators.argmap'> compilation 5", line 5, in argmap_write_graphml_lxml_1
  File "C:\Users\osthege\AppData\Local\Continuum\miniconda3\envs\pm3v4\lib\site-packages\networkx\readwrite\graphml.py", line 177, in write_graphml_lxml
    writer = GraphMLWriterLxml(
  File "C:\Users\osthege\AppData\Local\Continuum\miniconda3\envs\pm3v4\lib\site-packages\networkx\readwrite\graphml.py", line 746, in __init__
    self.add_graph_element(graph)
  File "C:\Users\osthege\AppData\Local\Continuum\miniconda3\envs\pm3v4\lib\site-packages\networkx\readwrite\graphml.py", line 787, in add_graph_element
    T = self.get_xml_type(self.attr_type(k, "node", v))
  File "C:\Users\osthege\AppData\Local\Continuum\miniconda3\envs\pm3v4\lib\site-packages\networkx\readwrite\graphml.py", line 457, in get_xml_type
    raise TypeError(
TypeError: GraphML does not support type <class 'type'> as data values.
```

I'm not familiar with `networkx`. Does the exception make any sense to you?

If you don't mind upgrading to `v4`, let's continue over in #5705 :)
Feel free to take over that branch (you'll have to open a new PR from your fork, or force-push away this one, because GitHub doesn't give you the permissions to push that branch in #5705).

**Comment by jonititan** [comment]
I'll have a look into transitioning my code to v4.   
Are there many breaking changes between v3 and v4?

I was just able to take advantage of your existing v3 export to graphviz feature to transfer the network into a networkx object.  
It should not have resulted in a breaking change to graphviz output unless there are code differences in how the export works in v3 and v4?

I'll review https://github.com/pymc-devs/pymc/blob/c22ea9689ece5a8d306852b5e50dbd99b515fb64/pymc/model_graph.py

**Comment by michaelosthege** [comment]
> I'll have a look into transitioning my code to v4. Are there many breaking changes between v3 and v4?

There are, but it depends how sophsticated your model is. You can check https://github.com/pymc-devs/pymc/blob/main/RELEASE-NOTES.md.

> I was just able to take advantage of your existing v3 export to graphviz feature to transfer the network into a networkx object. It should not have resulted in a breaking change to graphviz output unless there are code differences in how the export works in v3 and v4?

The Graphviz export is fine. I was trying to do `networkx.readwrite.write_graphml()` when I got that error.

**Comment by jonititan** [comment]
Yeah it looks like "type" might be a protected keyword for graphml?  Not a networkx issue per se?

**Comment by twiecki** [comment]
What's the benefit over graphviz plotting that we already have?

**Comment by jonititan** [comment]
I'm using NetworkX primarily for network analysis but also as an interim format which is easier to use with https://github.com/erdogant/d3graph

**Comment by michaelosthege** [comment]
> What's the benefit over graphviz plotting that we already have?

`networkx` can export to a broad range of other formats. This opens the door to a wider range of 2D and 3D visualization tools.

@ericmjl, you have graph experience. What do you think?

**Comment by ericmjl** [comment]
> > What's the benefit over graphviz plotting that we already have?
> 
> `networkx` can export to a broad range of other formats. This opens the door to a wider range of 2D and 3D visualization tools.
> 
> @ericmjl, you have graph experience. What do you think?

Agreed. Could be useful to use NetworkX graph objects as an intermediary.

**Comment by jonititan** [comment]
Unfortunately transitioning my code to v4 has been rather troublesome.
Especially since there does not appear to be v4 documentation as such currently.
I have managed to update some of the v3 examples to work under v4 but my own work seems to now interpret all constants as -inf which is a bit frustrating.

On the other hand I have rewritten model_graph.py and it does appear to work using v4.  
[model_graph.py](https://pastebin.com/x5XEBL5Z)

**Comment by ricardoV94** [comment]
> Unfortunately transitioning my code to v4 has been rather troublesome. 

Sorry to hear that. What trouble have you been finding?

> Especially since there does not appear to be v4 documentation as such currently.

The new docs are hidden in https://docs.pymc.io/en/latest/api.html

>  but my own work seems to now interpret all constants as -inf which is a bit frustrating.

What do you mean by this?

**Comment by jonititan** [comment]
I've revisited this and its now giving my TypeErrors.  (This is for a model that was working in v3)

[Error Message](https://pastebin.com/bGfqmK31)

I got rid of the -inf errors on constants by replacing them with very tightly bounded Gaussians but that isn't ideal.
My model is relatively simple in that it takes various distributions(and constants) as inputs to deterministic nodes which are pure functions.  These can then feed other deterministic nodes.  I'm modelling aircraft acquisition and operations costs.

**Comment by ricardoV94** [comment]
You seem to have type issues. We are a bit more strict about that in V4. 

Do you mind opening a topic on Discourse with reproducible code for us to assist you with? Difficult to gauge just with the error message (and this PR is not the right place for this anyway)

discourse.pymc.io/

Feel free to tag me over there.

**Comment by jonititan** [comment]
Will do.  I'll try to make a minimum reproducible example

**Comment by jonititan** [comment]
Done.  https://discourse.pymc.io/t/typeerror-when-combining-constant-and-normal-distribution-in-function-within-deterministic-node/9219
Should I close this PR since it is replaced by https://github.com/pymc-devs/pymc/pull/5705

**Comment by michaelosthege** [comment]
> Should I close this PR since it is replaced by https://github.com/pymc-devs/pymc/pull/5705

Whatever you prefer. More work is needed (addition of tests) and since you can't push the other branch, it might be more practical if you force-push it to this PR (`jonititan:patch-1`), or open a new PR with it.

**Pull request closed without merging**